streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java

```diff
@@ -192,6 +194,13 @@ private void determineIsQueryable(final String queryableStoreName) {
                            materializedInternal);
 
     }
+    
+    @Override
+    public <VR> CogroupedKStream<K, VR> cogroup(final Aggregator<? super K, ? super V, VR> aggregator) {
+        Objects.requireNonNull(aggregator, "aggregator should not be null");
+        return new CogroupedKStreamImpl<>(new KStreamBuilder(),
```

Comment:
This is the root cause of your error. You need to use one KStreamBuilder object -- because, you create a second one here, processors are added to different Topologies and thus the `join` in your test fails.

Reply:
@mjsax  Do you know where the first KStreamBuilder is instantiated? I looked through the code, but I could not find another KStreamBuilder constructor. From my understanding, only one instance of KStreamBuilder was involved throughout the entire process.

Reply:
I used a empty KStreamBuilder here after looking through Kyle's code. In the constructor, the cogroup() method was already called, so in other words, the initial <KGroupedStream, Aggregator> pair used in the constuctor was added to to the pairs field. If I insert, say, a KGroupedStream with a topology which already included the initial KGroupedStream, then we will have a duplicate set of source nodes in CogroupedKStream (remember it was added to pairs). Using a empty KStreamBuilder meanwhile prevents that issue.

Reply:
Try refactoring `CogroupedKStreamImpl` to take an `InternalStreamsBuilder` in the constructor instead of `KStreamBuilder`, just pass in the instance set by the constructor of the `KGroupedStreamImpl`.  Also, try removing `addSource` on line 182 of `CogroupedKStreamImpl`.  With those changes, I was able to get all but one of the tests to pass  I didn't have time to dig into the test failure, `KStreamCogroupIntegrationTest#testCogroupSessionWindow` but this should be a good start.

Reply:
The "original" `KStreamBuilder` is instantiated in the test code. Ie, the user create the object. The library should never create a `KStreamBuilder` or `StreamsBuilder` (or `InternalXXX` helper class). There must be one instance for each of those cases and those are all created when the user creates a `StreamsBuilder` -- from this point on, we just pass references around.